// @remotion/renderer/dist/options/image-format-option.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageFormatOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
let imageFormat = null;
const cliFlag = 'image-format';
exports.imageFormatOption = {
    name: 'Image Format',
    cliFlag,
    description: () => (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: ["Determines which in which image format to render. Can be either", ' ', jsx_runtime_1.jsx("code", { children: "\"jpeg\"" }),
            ", ",
            jsx_runtime_1.jsx("code", { children: "\"png\"" }),
            ",", ' ', jsx_runtime_1.jsx("code", { children: "\"webp\"" }),
            ", ",
            jsx_runtime_1.jsx("code", { children: "\"pdf\"" }),
            ", or", ' ', jsx_runtime_1.jsx("code", { children: "\"none\"" }),
            " depending on the render type."] })),
    ssrName: 'imageFormat',
    docLink: 'https://www.remotion.dev/docs/config#setvideoimageformat',
    type: null,
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return { value: commandLine[cliFlag], source: 'cli' };
        }
        if (imageFormat !== null) {
            return { value: imageFormat, source: 'config' };
        }
        return { value: null, source: 'default' };
    },
    setConfig(value) {
        imageFormat = value;
    },
};
