// @remotion/renderer/dist/options/private-license-key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.privateLicenseKeyOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const cliFlag = 'private-license-key';
let currentPrivateLicenseKey = null;
exports.privateLicenseKeyOption = {
    name: 'Private License Key',
    cliFlag,
    description: () => ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: ["The private license key for your company license, obtained from the \"Usage\" tab on ", (0, jsx_runtime_1.jsx)("a", { href: "https://remotion.pro/dashboard", children: "remotion.pro" }), ". If you are eligible for the free license, pass \"free-license\"."] })),
    ssrName: 'privateLicenseKey',
    docLink: 'https://www.remotion.dev/docs/licensing',
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return {
                source: 'cli',
                value: commandLine[cliFlag],
            };
        }
        if (currentPrivateLicenseKey !== null) {
            return {
                source: 'config',
                value: currentPrivateLicenseKey,
            };
        }
        return {
            source: 'default',
            value: null,
        };
    },
    setConfig: (value) => {
        currentPrivateLicenseKey = value;
    },
    type: null,
};
